core/lib/zksync_core/src/sync_layer/external_io.rs

```diff
@@ -459,10 +463,9 @@ impl StateKeeperIO for ExternalIO {
             panic!("State keeper requested to seal miniblock, but the next action is {action:?}");
         };
 
+        let store_latency = L1_BATCH_METRICS.start_storing_on_en();
         let mut storage = self.pool.access_storage_tagged("sync_layer").await.unwrap();
         let mut transaction = storage.start_transaction().await.unwrap();
-
-        let store_latency = L1_BATCH_METRICS.start_storing_on_en();
         // We don't store the transactions in the database until they're executed to not overcomplicate the state
         // recovery on restart. So we have to store them here.
         for tx in &updates_manager.miniblock.executed_transactions {
```

Comment:
EN currently saves transactions to DB in a sync manner (see lines https://github.com/matter-labs/zksync-era/pull/611/files#diff-31a0b1d8722a196212b88c601a2d540ee4be28a2bf32c534885e2910784c44baR471-R494). Options we have:
- Leave it as is, this part was taking only 40ms during today's high TPS period, 40ms/1000s per miniblock is pretty negligible. I prefer this approach
- Add a flag to `MiniblockSealCommand`, so it will insert transactions to DB if flag is true. Flag will be set to true for EN and false for main node. This a good option if saving will start taking more than 40ms.
- Change logic of `mark_txs_as_executed_in_miniblock`: currently it handles txs replacement with `UPDATE transactions` and we can change that to `INSERT .. ON CONFLICT DO UPDATE`, then we won't need to save transactions separately for EN. I don't really like it because on main node we will always fall into `ON CONFLICT` branch wasting time on trying to insert.

Reply:
Are you sure that we're not messing with EN correctness by splitting saving transactions and the miniblock data into separate transactions? The comment above implies that there may be complications with the state recovery on restart if transactions are saved in the DB before their execution; and if the EN crashes after saving transactions and before sealing a miniblock, that's what we'll end up with. IMO, adding a flag to `MiniblockSealCommand` to save transactions would be a safer option, and it (marginally) improves performance as well.

Reply:
I'm almost sure it will work. We do `INSERT .. ON CONFLICT DO NOTHING` and having those transactions in DB earlier shouldn't affect anything.
Though, I thought about it one more time and would prefer to be future-proof performance wise by adding flag to seal command. Will add this flag